go-libs/sqlexec/sqlexec.go

```diff
@@ -0,0 +1,263 @@
+		return nil
+	}
+	statusError := status.Error
+	if statusError == nil {
+		statusError = &sql.ServiceError{Message: "unknown", ErrorCode: sql.ServiceErrorCodeUnknown}
+	}
+	errorMessage := statusError.Message
+	if errorMessage == "" {
+		errorMessage = "unknown error"
+	}
+	if strings.Contains(errorMessage, "SCHEMA_NOT_FOUND") {
+		return fmt.Errorf("%s. %w", errorMessage, apierr.ErrNotFound)
+	}
+	if strings.Contains(errorMessage, "TABLE_OR_VIEW_NOT_FOUND") {
+		return fmt.Errorf("%s. %w", errorMessage, apierr.ErrNotFound)
+	}
+	// TODO: error mapping
+	return fmt.Errorf(errorMessage)
+}
+
+// Execute executes a SQL statement and returns the execution response
+func (s *StatementExecutionExt) Execute(ctx context.Context, warehouseID string, statement string, byteLimit int, catalog string, schema string, timeout time.Duration) (*sql.ExecuteStatementResponse, error) {
+	waitTimeout := ""
+	// if MinPlatformTimeout <= timeout.Seconds() && timeout.Seconds() <= MaxPlatformTimeout {
+	// 	waitTimeout = fmt.Sprintf("%fs", timeout.Seconds())
+	// }
+	logger.Infof(ctx, "Executing SQL statement: %s\n", statement)
+	immediateResponse, err := s.ExecuteStatement(ctx, sql.ExecuteStatementRequest{
+		WarehouseId: warehouseID,
+		Statement:   statement,
+		Catalog:     catalog,
+		Schema:      schema,
+		Disposition: sql.DispositionInline,
+		Format:      sql.FormatJsonArray,
+		ByteLimit:   int64(byteLimit),
+		WaitTimeout: waitTimeout,
+	})
+	if err != nil {
+		return nil, err
+	}
+	status := immediateResponse.Status
+	if status == nil {
+		status = &sql.StatementStatus{State: sql.StatementStateFailed}
+	}
+	if status.State == sql.StatementStateSucceeded {
+		return immediateResponse, nil
+	}
+	if err := s.raiseIfNeeded(status); err != nil {
+		return nil, err
+	}
+	attempt := 1
+	statusMessage := "polling..."
+	deadline := time.Now().Add(timeout)
+	statementID := immediateResponse.StatementId
+	for time.Now().Before(deadline) {
+		res, err := s.GetStatementByStatementId(ctx, statementID)
+		if err != nil {
+			return nil, err
+		}
+		resultStatus := res.Status
+		if resultStatus == nil {
+			return nil, fmt.Errorf("Result status is nil: %v", res)
+		}
+		state := resultStatus.State
+		if state == "" {
+			state = sql.StatementStateFailed
+		}
+		if state == sql.StatementStateSucceeded {
+			return &sql.ExecuteStatementResponse{
+				Manifest:    res.Manifest,
+				Result:      res.Result,
+				StatementId: statementID,
+				Status:      resultStatus,
+			}, nil
+		}
+		statusMessage = fmt.Sprintf("current status: %s", state)
+		if err := s.raiseIfNeeded(resultStatus); err != nil {
+			return nil, err
+		}
+		sleep := time.Duration(attempt)
+		if sleep > MaxSleepPerAttempt {
+			// sleep 10s max per attempt
+			sleep = MaxSleepPerAttempt
+		}
+		log.Printf("SQL statement %s: %s (sleeping ~%ds)\n", statementID, statusMessage, int(sleep.Seconds()))
```

Comment:
I would prefer to make it configurable, like, have this on DEBUG level...

Reply:
Good catch, it'll be debug - chatgpt didn't translate this part well